WIP: Add content type validation header

idanto opened the pull request:
Added header for header validation.
The only case that isn't covered is in case the body can be empty, in this case, the validation won't be smart and will fail if the header is missing.

I think this is an edge case but it's still work in progress, I will try to solve it.

idanto commented:
@igor-savin-ht you can review, If you have any good idea to handle the last open issue feel free to share.

idanto commented:
Basically, I think everything is ready for review, I solved all the issues by not setting the content-type header as required and using a regex for validation.
@igor-savin-ht @kibertoad can you please review?

kibertoad commented on src/middleware.js:
```diff
@@ -209,6 +215,18 @@ function buildInheritance(discriminator, dereferencedDefinitions, swagger, curre
+        'description': 'Allowed content types',
+        'required': true,
+        'pattern': 'application\/json.*\s*\S*',
+        // 'enum': contentTypes,
```
Is this needed?

kibertoad commented:
@idanto Would it be possible to skip validation in case body is empty? That sounds to be a more valid approach.

idanto commented:
@kibertoad It's much harder to implement, cause the whole idea behind ajv is the pre-compiled schema and currently the body and all other parameters are two different schemas.
I can check if there is a conditional check in Ajv and think how to indicate that there is something in the body but it will take me some time to solve

kibertoad commented:
@idanto Probably the easiest solution at this point is to make this check optional and disable by default, so that no existing requests start failing, and whoever needs this check, could opt-in.

idanto commented:
Please do not merge this one, I found an issue with the code - it is `application/json` specific and doesn't fully support the swagger spec.
 I would like also to make this optional as @kibertoad suggested

kibertoad commented:
@idanto Can we move the validation outside of ajv call and check the header manually?

idanto commented:
@kibertoad Please review again:
- Added content-type validation as an options
- Added validation only if `content-length` is higher than 0
- Added Support in multiple content-types (Supported by Swagger Spec)
- A small refactor to the code
- Added more tests to more edge cases

kibertoad commented on test/pet-store-swagger.yaml:
```diff
@@ -88,17 +88,20 @@ paths:
         - name: body
           in: body
-          required: true
+          # required: true
```
is this part of swagger supposed to be left out commented?

kibertoad commented on test/middleware-test.js:
```diff
@@ -1607,6 +1607,71 @@ describe('input-validation middleware tests', function () {
                     done();
                 });
         });
+        it('bad request - wrong content-type (should be application/json)', function (done) {
```
I presume that case when swagger specifies no consumed content type at all so it accepts whatever is covered by preexisting tests?

kibertoad replied:
Actually, it looks like used swagger defaults to application/json, so this case is most likely not covered. Would be nice to have.

kibertoad commented on src/middleware.js:
```diff
@@ -47,8 +51,15 @@ function init(swaggerPath, options) {
+
+                    if (bodySchema.length > 0) {
+                        schemas[parsedPath][currentMethod].body = buildBodyValidation(bodySchema[0].schema, dereferenced.definitions, swaggers[1], currentPath, currentMethod, parsedPath);
+                        // localParameters.push(createContentTypeHeaders(dereferenced.paths[currentPath][currentMethod].consumes || dereferenced.paths[currentPath].consumes || dereferenced.consumes));
```
should this still be there?

kibertoad commented on src/middleware.js:
```diff
@@ -2,10 +2,14 @@
+    InputValidationError = require('./inputValidationError'),
     schemaPreprocessor = require('./utils/schema-preprocessor');
 
+// require('ajv-keywords')(Ajv);
```
safe to remove?..

kibertoad commented on src/inputValidationError.js:
```diff
@@ -0,0 +1,74 @@
+            error.message = error.errors.message;
+        }
+
+        return error.dataPath + ' ' + error.message;
```
why not use template string?

kibertoad commented on src/customKeywords/content.js:
```diff
@@ -0,0 +1,31 @@
+const Ajv = require('ajv');
+
+module.exports = {
+    compile: function content(schema) {
```
contentTypeValidation would probably be more self-explanatory name?..

kibertoad commented on src/middleware.js:
```diff
@@ -47,8 +51,15 @@ function init(swaggerPath, options) {
+                        // localParameters.push(createContentTypeHeaders(dereferenced.paths[currentPath][currentMethod].consumes || dereferenced.paths[currentPath].consumes || dereferenced.consumes));
+                    }
+
+                    if (localParameters.length > 0 || middlewareOptions.contentTypeValidation) {
```
New configuration parameter doesn't seem to be documented.

kibertoad commented on README.md:
```diff
@@ -74,9 +74,9 @@ Options currently supports:
-
 - `ajvConfigBody` - Object that will be passed as config to new Ajv instance which will be used for validating request body. Can be useful to e. g. enable type coercion (to automatically convert strings to numbers etc). See Ajv documentation for supported values.
 - `ajvConfigParams` - Object that will be passed as config to new Ajv instance which will be used for validating request body. See Ajv documentation for supported values.
+- `contentTypeValidation` - Boolean that inticates if to perform content type validation in case `consume` field is specicy and the reqyest body is not empty.
```
"inticates" -> "indicates"
"specicy" -> "specified"
"reqyest" -> "request"

kibertoad commented:
@idanto Can I merge it?
